### DISCUSSION TITLE
Fabric v6 custom properties in serialized object (json)

### DISCUSSION THREAD

**Discussion by lance-lfx** [fabricjs/fabric.js#10176] [Q&A]
I'm having troubles trying to add custom properties to my serialized object. I have followed the demo in the new website to declare module 'fabric', typing the object with the custom properties in the interface and adding this line after: FabricObject.customProperties = ['name', 'id'];.

However, when I serialized the object, it is still not giving me those properties that I added. Any thoughts or recommendations?

**Comment by asturur** [comment]
Are you serializing a custom object or one that comes from fabric?

**Reply by lance-lfx** [reply]
I'm serializing a FabricObject that comes from fabric, not custom

**Reply by asturur** [reply]
Well then it should work or we have a bug.

The code concatenate the properties you ask, plus the fabricObject.customProperties and the specific object customProperties.

```
    const propertiesToSerialize = propertiesToInclude.concat(
      FabricObject.customProperties,
      (this.constructor as typeof FabricObject).customProperties || [],
    );
```

there is also a test:

```ts
  test('with default custom properties', () => {
    FabricObject.customProperties = ['key2'];
    const obj = new FabricObject({ name: 'test1', key2: 'value2' });
    expect(obj.name).toBe('test1');
    expect(obj.key2).toBe('value2');
    const serialized = obj.toObject();
    expect(serialized.name).toBe(undefined);
    expect(serialized.key2).toBe('value2');
  });
```

Is you code much different from this simple test?

**Reply by lance-lfx** [reply]
there are 2 differences from my code vs. the test you have shown above

1. i have the custom properties as optional properties, so they are typed like this in the interface: "id?: number"
2. instead of using the custom properties in a new instance, i was using FabricObject.set({ id: 123}).

the custom properties do show up in my object prior to serialization, but not after serialization

**Reply by asturur** [reply]
Please be clear what FabricObject is. Is it the variable name of you instance, or is it the imported class?
1. should not matter at all.

**Reply by lance-lfx** [reply]
it was a variable name of a FabricObject instance

**Reply by asturur** [reply]
customProperties is on the class, is not per instance.

**Comment by stewmorg** [comment]
I'm encountering this issue as well after following the guidelines in the documentation: https://fabricjs.com/docs/using-custom-properties/ 

my example is relatively simple and the IDE correctly detects the id and name properties set using the docs. However, when calling toObject or toJson the properties are not included in the serialisation. The documentation suggests it should be.

Am I missing a step?

``
const canvas = new StaticCanvas("canvasElementId");
    const helloWorld = new FabricText("hello world",{id:"123"});
    canvas.add(helloWorld);
    canvas.centerObject(helloWorld);
    console.log(helloWorld.id) // logs 123
    console.log(helloWorld.toJSON().id) // doesn't work
    console.log(canvas.toObject()); // no id property in the console.

**Reply by asturur** [reply]
<img width="685" alt="image" src="https://github.com/user-attachments/assets/b69e408a-2f68-4c57-856e-4fb419ea7aa5">

**Comment by brightchimp** [comment]
The [documentation](https://fabricjs.com/docs/using-custom-properties/) is slightly misleading...
```
// ** ADDED: This should go in a type definition file & is done at compilation **
import { FabricObject } from 'fabric';

declare module "fabric" {
  // to have the properties recognized on the instance and in the constructor
  interface FabricObject {
    id?: string;
    name?: string;
  }
  // to have the properties typed in the exported object
  interface SerializedObjectProps {
    id?: string;
    name?: string;
  }
}

// **ADDED: This should be executed at runtime so can go in a plugin file.**
// to actually have the properties added to the serialized object
FabricObject.customProperties = ['name', 'id'];
```

**Comment by mauricekindermann** [comment]
This feels like a bug to me so I made a quick demo:
https://jsfiddle.net/1b672vez

I'm trying to lock objects using native FabricJS properties (lockMovementX, selectable, etc.) via object.set(), but they don't appear in toJSON() output. The property exists on the object, just not in the serialized JSON.

```javascript
rect.set({ lockMovementX: true, selectable: false });

console.log(rect.lockMovementX);                       // true
console.log(canvas.toJSON().objects[0].lockMovementX); // undefined
```

**Reply by asturur** [reply]
the JSON by default capture the state of the visual canvas not the interactive part.
You can export those exactly as if they were custom properties.
